### PR TITLE
[processing] restore GDAL algorithms

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Restore and homogenize most of GDAL algorithms. Where possible extent parameter reprojected into input layer CRS, to avoid errors and freezes.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/plugins/processing/algs/gdal/ExecuteSql.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        self.dialects = ((self.tr('None'), ''),
+                         (self.tr('OGR SQL'), 'ogrsql'),
+                         (self.tr('SQLite'), 'sqlite'))
+
+        self.addParameter(QgsProcessingParameterFeatureSource(self.INPUT,
+                                                              self.tr('Input layer')))
+        self.addParameter(QgsProcessingParameterString(self.SQL,
+                                                       self.tr('SQL expression'),
+                                                       defaultValue=''))
+        self.addParameter(QgsProcessingParameterEnum(self.SQL,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/ExecuteSql.py]: Looks like a copy/paste mistake here?

File: python/plugins/processing/algs/gdal/GridAverage.py [hunk 2]
```diff
@@ -51,70 +53,122 @@ class GridAverage(GdalAlgorithm):
...
-                                          self.tr('Min points'), 0.0, 99999999.999999, 0.0))
-        self.addParameter(ParameterNumber(self.ANGLE,
-                                          self.tr('Angle'), 0.0, 359.0, 0.0))
-        self.addParameter(ParameterNumber(self.NODATA,
-                                          self.tr('Nodata'), 0.0, 99999999.999999, 0.0))
-        self.addParameter(ParameterSelection(self.RTYPE,
-                                             self.tr('Output raster type'), self.TYPE, 5))
-
-        self.addOutput(OutputRaster(self.OUTPUT, self.tr('Interpolated moving average')))
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridAverage.py]: Feature source here?

File: python/plugins/processing/algs/gdal/GridDataMetrics.py [hunk 3]
```diff
@@ -53,39 +55,87 @@ class GridDataMetrics(GdalAlgorithm):
...
-
-        self.addOutput(OutputRaster(self.OUTPUT, self.tr('Interpolated metrics')))
+        self.metrics = ((self.tr('Minimum'), 'minimum'),
+                        (self.tr('Maximum'), 'maximum'),
+                        (self.tr('Range'), 'range'),
+                        (self.tr('Count'), 'count'),
+                        (self.tr('Average distance'), 'average_distance'),
+                        (self.tr('Average distance between points'), 'average_distance_pts'))
+
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridDataMetrics.py]: Feature source?

File: python/plugins/processing/algs/gdal/GridInverseDistance.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    DATA_TYPE = 'DATA_TYPE'
+    OUTPUT = 'OUTPUT'
+
+    TYPES = ['Byte', 'Int16', 'UInt16', 'UInt32', 'Int32', 'Float32', 'Float64', 'CInt16', 'CInt32', 'CFloat32', 'CFloat64']
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridInverseDistance.py]: feature source?

File: python/plugins/processing/algs/gdal/GridInverseDistanceNearestNeighbor.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    DATA_TYPE = 'DATA_TYPE'
+    OUTPUT = 'OUTPUT'
+
+    TYPES = ['Byte', 'Int16', 'UInt16', 'UInt32', 'Int32', 'Float32', 'Float64', 'CInt16', 'CInt32', 'CFloat32', 'CFloat64']
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridInverseDistanceNearestNeighbor.py]: Feature source?

File: python/plugins/processing/algs/gdal/GridLinear.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    DATA_TYPE = 'DATA_TYPE'
+    OUTPUT = 'OUTPUT'
+
+    TYPES = ['Byte', 'Int16', 'UInt16', 'UInt32', 'Int32', 'Float32', 'Float64', 'CInt16', 'CInt32', 'CFloat32', 'CFloat64']
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridLinear.py]: Feature source

File: python/plugins/processing/algs/gdal/GridLinear.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        z_field_param = QgsProcessingParameterField(self.Z_FIELD,
+                                                    self.tr('Z value from field'),
+                                                    None,
+                                                    self.INPUT,
+                                                    QgsProcessingParameterField.Numeric,
+                                                    optional=True)
+        z_field_param.setFlags(z_field_param.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
+        self.addParameter(z_field_param)
+
+        self.addParameter(QgsProcessingParameterNumber(self.RADIUS,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridLinear.py]: What's a -1 search distance represent here?

File: python/plugins/processing/algs/gdal/GridNearestNeighbor.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    DATA_TYPE = 'DATA_TYPE'
+    OUTPUT = 'OUTPUT'
+
+    TYPES = ['Byte', 'Int16', 'UInt16', 'UInt32', 'Int32', 'Float32', 'Float64', 'CInt16', 'CInt32', 'CFloat32', 'CFloat64']
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterVectorLayer(self.INPUT,
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/GridNearestNeighbor.py]: feature source?

**Comment by nyalldawson** [comment]
Wow, fantastic work!

File: python/plugins/processing/algs/gdal/GridLinear.py (hunk 7 above)
Comment (alexbruy) [reply python/plugins/processing/algs/gdal/GridLinear.py]: From GDAL docs
>If set to -1, the search distance is infinite. If set to 0, nodata value will be always used.

File: python/plugins/processing/algs/gdal/ClipRasterByMask.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+    TYPES = ['Byte', 'Int16', 'UInt16', 'UInt32', 'Int32', 'Float32', 'Float64', 'CInt16', 'CInt32', 'CFloat32', 'CFloat64']
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterRasterLayer(self.INPUT,
+                                                            self.tr('Input layer')))
+        self.addParameter(QgsProcessingParameterVectorLayer(self.MASK,
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply python/plugins/processing/algs/gdal/ClipRasterByMask.py]: Hmm, not sure that for this algorithm it makes sense. I will investigate

**Comment by alexbruy** [comment]
Tests passed locally, but failed on Travis (probably because different GDAL versions/builds are used).

**Comment by nirvn** [comment]
@alexbruy , anything holding this nice PR off?

**Comment by alexbruy** [comment]
@nirvn, only failing tests. They all pass for me locally, but fail on Travis. I just rebased to have latest changes (using docker for testing) and still no luck.

**Comment by alexbruy** [comment]
I disabled tests which failed because of missed GDAL Python scrips (e.g. `gdal2xyz.py`, `gdal_polygonize.py` etc) and now all test passed. The only "failing" test is spellcheck, it timed out probably because of massive update of expected files. IMHO it is safe to ignore this failure and merge.

@nyalldawson, @m-kuhn are you ok with merging this despite failing spellcheck test?

**Comment by nyalldawson** [comment]
No problem merging with spell check timeout... It'll fix itself after merge.

Great work on this!!

**Comment by nirvn** [comment]
Woupidou!

**Pull request merged**